app/src/org/commcare/activities/connect/ConnectIDPhoneVerificationActivity.java

```diff
@@ -0,0 +1,281 @@
+
+    private final Handler taskHandler = new android.os.Handler();
+
+    private final Runnable runnable = new Runnable() {
+        @Override
+        public void run() {
+            if (smsTime != null) {
+                double elapsedMinutes = ((new DateTime()).getMillis() - smsTime.getMillis()) / 60000.0;
+                int resendLimitMinutes = 2;
+                boolean allowResend = elapsedMinutes > resendLimitMinutes;
+                uiController.setResendEnabled(allowResend);
+
+                String text = getString(R.string.connect_verify_phone_resend);
+                if (!allowResend) {
+                    text = getString(R.string.connect_verify_phone_resend_wait, (int) Math.ceil((resendLimitMinutes - elapsedMinutes) * 60));
+                }
+                else {
+                    smsTime = null;
+                }
+
+                uiController.setResendText(text);
+            }
+
+            taskHandler.postDelayed(this, 100);
+        }
+    };
+
+    void startHandler() {
+        taskHandler.postDelayed(runnable, 100);
+    }
+
+    void stopHandler() {
+        taskHandler.removeCallbacks(runnable);
+    }
+
+    public void updateMessage() {
+        boolean alternate = method == MethodRecoveryAlternate;
+        String phone = alternate ? recoveryPhone : primaryPhone;
+        if(phone == null) {
+            phone = "-";
+        }
+
+        uiController.setLabelText(getString(R.string.connect_verify_phone_label, phone));
+    }
+
+    public void requestSMSCode() {
+        smsTime = new DateTime();
+
+        uiController.setErrorMessage(null);
+        String command;
+        HashMap<String, String> params = new HashMap<>();
+        AuthInfo authInfo = new AuthInfo.NoAuth();
+        switch(method) {
+            case MethodRecoveryPrimary -> {
+                command = "/users/recover";
+                params.put("phone", username);
+            }
+            case MethodRecoveryAlternate -> {
+                command = "/users/recover/secondary";
+                params.put("phone", username);
+                params.put("secret_key", password);
+            }
+            default -> {
+                command = "/users/validate_phone";
+                authInfo = new AuthInfo.ProvidedAuth(username, password, false);
+            }
+        }
+        String url = getString(R.string.ConnectURL) + command;
+
+        ConnectIDNetworkHelper.post(this, url, authInfo, params, false, new ConnectIDNetworkHelper.INetworkResultHandler() {
+            @Override
+            public void processSuccess(int responseCode, InputStream responseData) {
+                try {
+                    String responseAsString = new String(StreamsUtil.inputStreamToByteArray(responseData));
+                    if(responseAsString.length() > 0) {
+                        JSONObject json = new JSONObject(responseAsString);
+                        String key = "secret";
```

Comment:
any json keys should be defined as constants.

Reply:
Done.